### PR TITLE
121 rigid body rod contact friction

### LINKED ISSUE #121: Rigid Body - Rod contact friction
**Is your feature request related to a problem? Please describe.**
We need to add a friction model between rod and cylinder that are in contact. Currently there is only contact force between the two object.

**Describe the solution you'd like**
Rod and cylinder should not slip pass each other for large friction coefficients.

### PR DESCRIPTION
**Pull request by armantekinalp**
Fixes #121 We implemented friction model between rod and cylinder that are in contact, together with example cases. 

Here are the some of the results. `rod_one` is the Cosserat Rod, `rod_two` is the cylinder. We apply a force on `rod_one` and fix the position of `cylinder`. For the below case there is no friction between them.

rod_rigid_velocity

https://user-images.githubusercontent.com/53585636/177080428-143b8fea-f1fa-479c-9643-55e582929a75.mp4

In the second case we added friction and `rod_one` does not move. 

rod_rigid_velocity

https://user-images.githubusercontent.com/53585636/177080467-a7edfd73-2a3f-472e-9df8-1fd15bee0b71.mp4

Also we varied the the applied force on `rod_one` and plotted the energy vs applied external force. As it can be seen from the below graph after applied force is greater than some value, energy of `rod_one` starts to increase. In other words `rod_one` starts to move.

rod_energy_vs_force

### CODE REVIEW

**Comment by armantekinalp** [comment]
@poratamir I am adding you to this PR, since you are interested in rod-cylinder contact.

**Comment by poratamir** [comment]
Thank you!

File: elastica/joint.py [hunk 1]
```diff
@@ -771,20 +803,66 @@ def _prune_using_aabbs_rod_rod(
...
+    Examples
+    --------
+    How to define contact between rod and cylinder.
+
+    >>> simulator.connect(rod, cylidner).using(
+    ...    ExternalContact,
+    ...    k=1e4,
+    ...    nu=10,
+    ...    velocity_damping_coefficient=10,
+    ...    kinetic_friction_coefficient=10,
```
Comment (bhosale2) [inline elastica/joint.py:823]: show illustration range value in 0-1 range maybe?

File: elastica/joint.py [hunk 2]
```diff
@@ -771,20 +803,66 @@ def _prune_using_aabbs_rod_rod(
...
+
+    Examples
+    --------
+    How to define contact between rod and cylinder.
+
+    >>> simulator.connect(rod, cylidner).using(
+    ...    ExternalContact,
+    ...    k=1e4,
+    ...    nu=10,
+    ...    velocity_damping_coefficient=10,
```
Comment (bhosale2) [inline elastica/joint.py:822]: maybe this one should be set higher to keep slip threshold smaller? maybe we can add the following to the notes:

"The `velocity_damping_coefficient` is set to a high value (e.g. 1e4) to minimize slip and simulate stiction (static friction), while `friction_coefficient` corresponds to the Coulombic friction coefficient."

File: examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        base_length,
+        base_radius,
+        density,
+        nu,
+        E,
+        shear_modulus=shear_modulus,
+    )
+
+    rod_cylinder_parallel_contact_simulator.append(rod)
+    # rod.velocity_collection[0, :] -= 0.2
```
Comment (bhosale2) [inline examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py:50]: remove?

File: examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    # Apply uniform forces on the rod
+    rod_cylinder_parallel_contact_simulator.add_forcing_to(rod).using(
+        UniformForces, force=1.0 * pulling_force, direction=direction
+    )
+    # Push the rod towards the cylinder to make sure contact is there
+    rod_cylinder_parallel_contact_simulator.add_forcing_to(rod).using(
+        UniformForces, force=0.1, direction=np.array([-1.0, 0.0, 0.0])
```
Comment (bhosale2) [inline examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py:58]: define a variable as `normal_force` and set `force=normal_force`?

File: examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+        E,
+        shear_modulus=shear_modulus,
+    )
+
+    rod_cylinder_parallel_contact_simulator.append(rod)
+    # rod.velocity_collection[0, :] -= 0.2
+
+    # Apply uniform forces on the rod
+    rod_cylinder_parallel_contact_simulator.add_forcing_to(rod).using(
+        UniformForces, force=1.0 * pulling_force, direction=direction
```
Comment (bhosale2) [inline examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py:54]: set `pulling_force = gamma * normal_force`? then as gamma crosses coeff_friction the rod will move?

File: examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        OneEndFixedBC, constrained_position_idx=(0,), constrained_director_idx=(0,)
+    )
+
+    # Add contact between rigid body and rod
+    rod_cylinder_parallel_contact_simulator.connect(rod, rigid_body).using(
+        ExternalContact,
+        k=1e4,
+        nu=10,
+        velocity_damping_coefficient=10,
+        kinetic_friction_coefficient=10,
```
Comment (bhosale2) [inline examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py:89]: set first one to 1e3-1e4 and second to 0-1 range ?

File: examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+if __name__ == "__main__":
+    import multiprocessing as mp
+    from examples.RigidBodyCases.RodRigidBodyContact.post_processing import (
+        plot_force_vs_energy,
+    )
+
+    # total_energy = rod_cylinder_contact_friction_case(pulling_force=0.5, POST_PROCESSING=True)
+
+    force = list([(float(x)) / 100.0 for x in range(0, 90, 5)])
```
Comment (bhosale2) [inline examples/RigidBodyCases/RodRigidBodyContact/rod_cylinder_contact_friction.py:272]: here we can then fix normal_force and iterate over pulling_force = gamma * normal_force with gamma in a range from 0 to 1 ?

**Review by bhosale2 — CHANGES_REQUESTED** [review]
Minor changes, @armantekinalp have a look and see if they make sense, I can help out in implementing these changes.

**Comment by armantekinalp** [comment]
Modified phase space.

rod_energy_vs_force

**Review by bhosale2 — APPROVED** [review]
LGTM.

**Pull request merged**